### PR TITLE
Call ocr validation retry

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-709

### Change description ###
Replacing PR-1551

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuite.java [hunk 1]
```diff
@@ -125,7 +125,7 @@
     protected BlobContainerClient rejectedContainer;
 
     private static DockerComposeContainer dockerComposeContainer;
-    private static String dockerHost;
+    static String dockerHost;
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuite.java:128]: ?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuiteForOcrValidation.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            HttpHeaders.EMPTY,
+            null,
+            null
+        );
+    }
+
+    private void assertNoEnvelopesInDb() {
+        // We expect only one envelope which was uploaded
+        List<Envelope> envelopes = envelopeRepository.findAll();
+        assertThat(envelopes).hasSize(0);
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuiteForOcrValidation.java:76]: comment says it expects 1

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java [hunk 3]
```diff
@@ -96,6 +98,19 @@ public void processZipFileContent(
         } catch (PreviouslyFailedToUploadException ex) {
             log.warn("Rejected file {} from container {} - failed previously", zipFilename, containerName, ex);
             createEvent(DOC_UPLOAD_FAILURE, containerName, zipFilename, ex.getMessage());
+        } catch (OcrValidationServerSideException ex) {
+            log.error("Failed to process file {} from container {}, "
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java:102]: new line

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java [hunk 4]
```diff
@@ -85,12 +91,44 @@ public OcrValidator(
...
+    private OcrValidationWarnings handleOcrValidationError(
+        Throwable ex,
+        BlobClient blobClient,
+        String leaseId,
+        InputScannableItem docWithOcr,
+        String validationUrl,
+        String zipFileName
+    ) {
+        if (ex instanceof HttpServerErrorException) {
+            if (ocrValidationRetryManager.setRetryDelayIfPossible(blobClient, leaseId)) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java:114]: shouldn't this be one `if`?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForOcrValidationFailedOnServerSide.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            .willThrow(getServerSideException())
+            .willThrow(getServerSideException())
+        ;
+
+        // when
+        processor.processBlobs();
+
+        retryAfterDelay();
+
+        retryAfterDelay();
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForOcrValidationFailedOnServerSide.java:46]: why twice? please add a comment in code

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuite.java (hunk 1 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuite.java:128]: Made private

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuiteForOcrValidation.java (hunk 2 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ProcessorTestSuiteForOcrValidation.java:76]: Fixed

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java (hunk 3 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java:102]: What do you mean? @rkondratowicz

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java (hunk 4 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java:114]: Will do

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java (hunk 3 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java:102]: first parameter should be in new line

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/EnvelopeHandlerTest.java [hunk 6]
```diff
@@ -144,7 +145,47 @@ void should_handle_and_save_envelope() {
         assertThat(envelope.getValue().getPayments()).isEqualTo(inputEnvelope.payments);
         assertThat(envelope.getValue().getNonScannableItems()).isEqualTo(inputEnvelope.nonScannableItems);
         assertThat(envelope.getValue().getContainer()).isEqualTo(CONTAINER_NAME);
-        assertThat(envelope.getValue().getRescanFor()).isEqualTo(RESCAN_FOR_FILE_NAME);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/EnvelopeHandlerTest.java:147]: why was this removed?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java [hunk 7]
```diff
@@ -96,6 +98,20 @@ public void processZipFileContent(
...
+                    + "Error message: {}, status code: {}, status text: {}"
+                    + " - will retry",
+                zipFilename,
+                containerName,
+                ex.getMessage(),
+                ex.getCause().getMessage(),
+                ((HttpServerErrorException) ex.getCause()).getStatusCode().toString(),
+                ((HttpServerErrorException) ex.getCause()).getStatusText(),
+                ex
+            );
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/FileContentProcessor.java:114]: all other exception result in an event. should this one not do this as well?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java [hunk 8]
```diff
@@ -85,12 +91,43 @@ public OcrValidator(
...
-                        )
-                    )
+                    .getOrElseGet(exc -> handleOcrValidationError(
+                        exc,
+                        blobClient,
+                        leaseId,
+                        docWithOcr,
+                        validationUrl,
+                        envelope.zipFileName
+                    ))
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java:101]: aren't http exceptions handled in `.onFailure`? (line 92)
